### PR TITLE
Support Formatter on Legend

### LINKED ISSUE #2125: Enhancement Request: Formatter on legend
I want to make it obvious these are percents
screen shot 2015-05-20 at 4 11 58 pm

### PR DESCRIPTION
**Pull request by yunhaolucky**
**Feature**
Display a different name on an ordinary Legend than the values in the colorScale domain using customized `Formatter`.`Formatter` is defined as `(d:any) => string`
```typescript
class Legend{
/**
* Gets the Formatter for the text.
*/
formatter(): Formatter;
/**
* Sets the Formatter for the text.
*
* @param {Formatter} formatter
* @returns {Legend} The calling Legend.
*/
formatter(formatter: Formatter): Legend;
}
```

Using  `formatter` will affect the behavior of `comparator`. We decide to make `comparator` work on formatted texts.

**JSFiddle**
Before:http://jsfiddle.net/yunhaolucky/cmf3ut7g/4/
After:http://jsfiddle.net/yunhaolucky/cmf3ut7g/3/

close #2125

### CODE REVIEW

File: src/components/interpolatedColorLegend.ts [hunk 1]
```diff
@@ -68,7 +68,7 @@ export module Components {
      */
     public formatter(formatter: Formatter): InterpolatedColorLegend;
     public formatter(formatter?: Formatter): any {
-      if (formatter === undefined) {
+      if (formatter === null) {
```
Comment (yunhaolucky) [inline src/components/interpolatedColorLegend.ts]: This is bug in `interpolatedColorLegend`, not part of this feature.

Comment (bluong) [inline src/components/interpolatedColorLegend.ts]: Can you explain the bug?  The way that you have currently implemented this change completely changes the functionality of this function.  Essentially, I believe the getter is now called with `icl.formater(null);` which seems very undesirable.

Comment (bluong) [inline src/components/interpolatedColorLegend.ts]: If this is a bug fix for `InterpolatedColorLegend` that seems to be separate from this pull request, please move it out into a separate pull request.

File: src/components/legend.ts [hunk 2]
```diff
@@ -63,6 +64,25 @@ export module Components {
...
+    public formatter(): Formatter;
+    /**
+     * Sets the Formatter for the text.
+     *
+     * @param {Formatter} formatter
+     * @returns {Legend} The calling Legend.
+     */
+    public formatter(formatter: Formatter): Legend;
+    public formatter(formatter?: Formatter): any {
+      if (formatter === null) {
```
Comment (bluong) [inline src/components/legend.ts]: This seems incorrect.  Now the getter can only be called with `legend.formatter(null)` it seems.

File: test/components/legendTests.ts [hunk 3]
```diff
@@ -254,6 +254,34 @@ describe("Legend", () => {
     svg.remove();
   });
 
+  it("accept formatter and show entity labels correcly", () => {
+    color.domain(["jtao", "mschafer", "kfalter"]);
+    let data = [
+      {id: "jtao", name: "Joy Tao", score: 13},
+      {id: "mschafer", name: "Mark Schafer", score: 15},
+      {id: "kfalter", name: "Kelsey Falter", score: 11}
+    ];
```
Comment (bluong) [inline test/components/legendTests.ts]: I personally am find with using first names of people but unsure about using full names of people.  Would it make sense to just use first names?

File: test/components/legendTests.ts [hunk 4]
```diff
@@ -254,6 +254,34 @@ describe("Legend", () => {
...
+    let f = (id: string) => {
```
Comment (bluong) [inline test/components/legendTests.ts]: `let formatter = ...`

File: test/components/legendTests.ts [hunk 5]
```diff
@@ -254,6 +254,34 @@ describe("Legend", () => {
     svg.remove();
   });
 
+  it("accept formatter and show entity labels correcly", () => {
```
Comment (bluong) [inline test/components/legendTests.ts]: how about
`it("can set which formatter to use to change how entry labels are displayed"...`?

File: test/components/legendTests.ts [hunk 6]
```diff
@@ -254,6 +254,34 @@ describe("Legend", () => {
...
+      {id: "mschafer", name: "Mark Schafer", score: 15},
+      {id: "kfalter", name: "Kelsey Falter", score: 11}
+    ];
+    let f = (id: string) => {
+      let pickedEntity = data.filter(function(d) {return (d.id === id); });
+      if (pickedEntity.length > 0) {
+          return pickedEntity[0].name;
+        }else {
+          return "NA";
+        }
```
Comment (bluong) [inline test/components/legendTests.ts]: Indentation issues above.

Comment (bluong) [inline test/components/legendTests.ts]: Instead of the logic above, why can't we create a map instead of a data array?  It'd be immensely easier to work with here.

Just noting that it seems you're adding logic to this formatter that seems unnecessary (this "NA" case, the `pickedEntity.length > 0` case), which can make deciphering this test harder to understand when we look at this later.

File: test/components/legendTests.ts [hunk 7]
```diff
@@ -254,6 +254,34 @@ describe("Legend", () => {
...
+    };
+    legend.formatter(f);
+    legend.renderTo(svg);
+
+    let texts = svg.selectAll("title")[0].map(function(text: Element) {return text.textContent; });
```
Comment (bluong) [inline test/components/legendTests.ts]: Why are we looking for all of the `title`'s on the svg?  While I think this is one way we can check the formatter, a lot more useful check I think is to check for the text inside each entry.

Comment (bluong) [inline test/components/legendTests.ts]: I believe D3 has a `text()` API that can be incredibly useful.

https://github.com/mbostock/d3/wiki/Selections#text

File: test/components/legendTests.ts [hunk 8]
```diff
@@ -254,6 +254,34 @@ describe("Legend", () => {
...
+    data.forEach(function(d, i) {
+      let idExist = (texts.indexOf(d.id) !== -1);
+      let nameExist = (texts.indexOf(d.name) !== -1);
+      assert.isTrue(nameExist, `formatter output ${d.name} should be drawn`);
+      assert.isFalse(idExist, ` formatter input ${d.id} should not be drawn`);
+    });
+     svg.remove();
```
Comment (bluong) [inline test/components/legendTests.ts]: Weird indentation issues.

File: test/components/legendTests.ts [hunk 9]
```diff
@@ -254,6 +254,34 @@ describe("Legend", () => {
...
+    };
+    legend.formatter(f);
+    legend.renderTo(svg);
+
+    let texts = svg.selectAll("title")[0].map(function(text: Element) {return text.textContent; });
+    data.forEach(function(d, i) {
+      let idExist = (texts.indexOf(d.id) !== -1);
+      let nameExist = (texts.indexOf(d.name) !== -1);
+      assert.isTrue(nameExist, `formatter output ${d.name} should be drawn`);
+      assert.isFalse(idExist, ` formatter input ${d.id} should not be drawn`);
```
Comment (bluong) [inline test/components/legendTests.ts]: I think checking the latter is not important.  I think we only care that the formatter works

File: test/components/legendTests.ts (hunk 3 above)
Comment (bluong) [inline test/components/legendTests.ts]: The `score` parameter is not used anywhere.  I think we can just delete this parameter

File: test/components/legendTests.ts [hunk 10]
```diff
@@ -254,6 +254,34 @@ describe("Legend", () => {
...
+        }
+    };
+    legend.formatter(f);
+    legend.renderTo(svg);
+
+    let texts = svg.selectAll("title")[0].map(function(text: Element) {return text.textContent; });
+    data.forEach(function(d, i) {
+      let idExist = (texts.indexOf(d.id) !== -1);
+      let nameExist = (texts.indexOf(d.name) !== -1);
+      assert.isTrue(nameExist, `formatter output ${d.name} should be drawn`);
```
Comment (bluong) [inline test/components/legendTests.ts]: Instead of using `assert.isTrue` can we use something that's more useful?  Maybe loop over the appropriate domain value and apply the locally defined formatter to it?

File: test/components/legendTests.ts (hunk 8 above)
Comment (bluong) [inline test/components/legendTests.ts]: There's no test above for the `getter` for the formatter.  This is incredibly important since I have a suspicion that the getter does not work.

File: plottable.d.ts [hunk 11]
```diff
@@ -2160,6 +2160,17 @@ declare module Plottable {
             constructor(colorScale: Scales.Color);
             protected _setup(): void;
             /**
+             * Gets the Formatter for the text.
+             */
+            formatter(): Formatter;
+            /**
+             * Sets the Formatter for the text.
```
Comment (bluong) [inline plottable.d.ts]: Hmm maybe be more specific to which text you're referring?  Maybe
"Sets the Formatter for the entries (or entry texts)."?  I would ping to get a second opinion

File: src/components/legend.ts [hunk 12]
```diff
@@ -144,13 +164,17 @@ export module Components {
...
       let availableWidthForEntries = Math.max(0, (availableWidth - this._padding));
 
-      let entryNames = this._colorScale.domain().slice();
-      entryNames.sort(this.comparator());
+      let entryLabels: d3.Map<string> = d3.map<string>();
+      this._colorScale.domain().slice().forEach((d) => {
+        entryLabels.set(this._formatter(d), d);
+      });
+      let entryNames = entryLabels.keys();
+      entryNames = entryNames.sort(this.comparator()).map((d) => entryLabels.get(d));
```
Comment (bluong) [inline src/components/legend.ts]: using a map here seems very roundabout.  Why not `entryNames = entryNames.sort(this.comparator()).map(this.formatter())`?

File: test/components/legendTests.ts [hunk 13]
```diff
@@ -254,6 +254,35 @@ describe("Legend", () => {
...
+    color.domain(["jtao", "mschafer", "kfalter"]);
+    let data: {[id: string]: string; } = {"jtao": "Joy", "mschafer": "Mark", "kfalter": "Kelsey"};
+    let formatter = (id: string) => {
+        return data[id];
+    };
+    legend.formatter(formatter);
+    legend.renderTo(svg);
+    assert.strictEqual(legend.formatter(), formatter, "formatter() return the formatter of legend correctly");
+    svg.remove();
+  });
```
Comment (bluong) [inline test/components/legendTests.ts]: the `renderTo` line above does not really do anything.  Remove it?

File: src/components/legend.ts (hunk 12 above)
Comment (yunhaolucky) [inline src/components/legend.ts]: I want to sort the labels by `this.formatter(entryID)` instead of `entryID`.  Is there any good way to do something like `entryNames.sort(this.comparator(this.formatter()))`?

Comment (bluong) [inline src/components/legend.ts]: Please check if this is what people want first. If this is indeed the case, you can map the formatter over all of the labels and then apply the sort then

`entryNames.map(this.formatter()).sort(this.comparator())`

**Comment by bluong** [comment]
Please check what people want out of this

File: src/components/legend.ts (hunk 12 above)
Comment (yunhaolucky) [inline src/components/legend.ts]: The reason I did this is I want to have a array of scale domain names sorted by displayed texts. If I do `entryNames.map(this.formatter()).sort(this.comparator())`, I will not get the domain names back. I need domain names to pick the color in ColorScale.

File: test/components/legendTests.ts [hunk 14]
```diff
@@ -254,6 +254,46 @@ describe("Legend", () => {
...
+    let data: {[id: string]: string; } = {"jtao": "Joy", "mschafer":  "Mark", "kfalter": "Kelsey"};
+    let formatter = (id: string) => data[id];
+    legend.formatter(formatter);
+    legend.renderTo(svg);
+    let legendRows = svg.selectAll(entrySelector);
+    assert.operator(legendRows.size(), ">=", 1, "There is at least one entry in the test");
+    legendRows.each(function(d: Element, i: number){
+      let expectText = formatter(legend.colorScale().domain()[i]);
+      assert.strictEqual(d3.select(this).select("text").text(), expectText,
+        `formatter output ${d3.select(this).select("text").text()} should be displayed`);
```
Comment (bluong) [inline test/components/legendTests.ts]: Shouldn't the test message include `expectText` instead of the actual response?

File: src/components/legend.ts (hunk 12 above)
Comment (bluong) [inline src/components/legend.ts]: > I need domain names to pick the color in ColorScale.

What do you mean by this?

Comment (bluong) [inline src/components/legend.ts]: The variable names here can be quite a bit confusing.

It seems that `entryNames` is trying to be a variable for 2 different concepts.  This can be bad, because if one of the 2 concepts needs to change, then the variable here has to be drastically modified.  One is that it holds the formatted transform of the colorScale's domain.  Two is that is also eventually holds the original colorScale's domain in sorted form.

Can we separate these two variables out and see how it reads?

File: src/components/legend.ts [hunk 15]
```diff
@@ -144,13 +164,17 @@ export module Components {
 
       let availableWidthForEntries = Math.max(0, (availableWidth - this._padding));
 
-      let entryNames = this._colorScale.domain().slice();
-      entryNames.sort(this.comparator());
+      let entryLabels: d3.Map<string> = d3.map<string>();
+      this._colorScale.domain().slice().forEach((d) => {
+        entryLabels.set(this._formatter(d), d);
```
Comment (bluong) [inline src/components/legend.ts]: Curious why this isn't `entryLabels.set(d, this._formatter(d))`.  One usually has a mapping from a raw value to a computed value, not usually the other way around.

**Comment by bluong** [comment]
Semantics semantics

File: src/components/legend.ts (hunk 15 above)
Comment (yunhaolucky) [inline src/components/legend.ts]: To get computed value from raw value, I can use function formatter(value). In this case, I want to get raw value from computed value. That's why I want to use a map to store those values.

File: src/components/legend.ts (hunk 12 above)
Comment (yunhaolucky) [inline src/components/legend.ts]: >What do you mean by this?

 We use `this._colorScale.scale(domainNames)` to get the color of a entry symbol. And formatter can only translate from domain names to displayed texts. Therefore, I need to store domain names somewhere. That's the reason I want an array of domain names sorted by its corresponding displayed texts instead of an array of sorted displayed texts.

File: test/components/legendTests.ts [hunk 16]
```diff
@@ -254,6 +254,46 @@ describe("Legend", () => {
...
+      let expectText = formatter(legend.colorScale().domain()[i]);
+      assert.strictEqual(d3.select(this).select("text").text(), expectText,
+        `formatter output ${expectText} should be displayed`);
+    });
+    svg.remove();
+  });
+
+  it("can get formatter of the legend using formatter()", () => {
+    color.domain(["jtao", "mschafer", "kfalter"]);
+    let data: {[id: string]: string; } = {"jtao": "Joy", "mschafer": "Mark", "kfalter": "Kelsey"};
```
Comment (bluong) [inline test/components/legendTests.ts]: The above two lines are not too relevant

File: test/components/legendTests.ts [hunk 17]
```diff
@@ -254,6 +254,46 @@ describe("Legend", () => {
...
+    let formatter = (id: string) => data[id];
```
Comment (bluong) [inline test/components/legendTests.ts]: We can just use a very simple formatter here

File: test/components/legendTests.ts [hunk 18]
```diff
@@ -254,6 +254,46 @@ describe("Legend", () => {
...
+    color.domain(colorDomain);
+    let expectedTexts = ["Z", "Y", "X"];
+    let formatter = (d: string) => expectedTexts[colorDomain.indexOf(d)];
+    legend.formatter(formatter);
+    let comparator = (a: string, b: string) => a.localeCompare(b);
+    legend.comparator(comparator);
+    legend.renderTo(svg);
+    let entryTexts = svg.selectAll(entrySelector)[0].map((node: Element) => d3.select(node).select("text").text());
+    expectedTexts.sort(comparator);
+    assert.deepEqual(expectedTexts, entryTexts, "formatted entries have been sorted as expected");
```
Comment (bluong) [inline test/components/legendTests.ts]: what does "sorted as expected" mean?

File: test/components/legendTests.ts [hunk 19]
```diff
@@ -254,6 +254,46 @@ describe("Legend", () => {
...
+  it("can get formatter of the legend using formatter()", () => {
+    color.domain(["jtao", "mschafer", "kfalter"]);
+    let data: {[id: string]: string; } = {"jtao": "Joy", "mschafer": "Mark", "kfalter": "Kelsey"};
+    let formatter = (id: string) => data[id];
+    legend.formatter(formatter);
+    assert.strictEqual(legend.formatter(), formatter, "formatter() returns the formatter of legend correctly");
+    svg.remove();
+  });
+
+  it("comparator() works as expected when customized formatter is applied", () => {
```
Comment (bluong) [inline test/components/legendTests.ts]: Should comment that it sorts the formatted text

File: test/components/legendTests.ts [hunk 20]
```diff
@@ -254,6 +254,46 @@ describe("Legend", () => {
...
+    let colorDomain = ["A", "B", "C"];
+    color.domain(colorDomain);
+    let expectedTexts = ["Z", "Y", "X"];
+    let formatter = (d: string) => expectedTexts[colorDomain.indexOf(d)];
+    legend.formatter(formatter);
+    let comparator = (a: string, b: string) => a.localeCompare(b);
```
Comment (bluong) [inline test/components/legendTests.ts]: Why don't we use a more simple comparator then `a.localeCompare(b)`

File: test/components/legendTests.ts [hunk 21]
```diff
@@ -254,6 +254,46 @@ describe("Legend", () => {
...
+    legend.comparator(comparator);
+    legend.renderTo(svg);
+    let entryTexts = svg.selectAll(entrySelector)[0].map((node: Element) => d3.select(node).select("text").text());
+    expectedTexts.sort(comparator);
```
Comment (bluong) [inline test/components/legendTests.ts]: Why do we have to sort the `expectedTexts` as well?

File: src/components/legend.ts [hunk 22]
```diff
@@ -144,13 +164,16 @@ export module Components {
...
 
-      let entryNames = this._colorScale.domain().slice();
-      entryNames.sort(this.comparator());
-
+      let entryLabels: d3.Map<string> = d3.map<string>();
+      this._colorScale.domain().slice().forEach((d) => {
+        entryLabels.set(this._formatter(d), d);
+      });
+      let formattedEntryNames = entryLabels.keys();
+      let entryNames = formattedEntryNames.sort(this.comparator()).map((d) => entryLabels.get(d));
```
Comment (bluong) [inline src/components/legend.ts]: Would this work?
`let entryNames = this._colorScale.domain().sort((a, b) => this.comparator()(this.formatter()(a), this.formatter()(b));`

This would get rid of the the `entryLabels` and the `formattedEntryNames` variable and I believe this is still readable

**Comment by bluong** [comment]
The unit tests can be made quite a bit more simpler / easier to understand.

Also we can tighten the logic of the source code by quite a bit.

File: test/components/legendTests.ts (hunk 21 above)
Comment (yunhaolucky) [inline test/components/legendTests.ts]: My logic is:
Given a sorted array of domain names and an unsorted array of expected texts, the actual displayed texts should be sorted. So I need to compare the actual displayed texts and sorted expected texts to see whether comparator sorts in the order of displayed texts.

File: src/components/legend.ts [hunk 23]
```diff
@@ -45,9 +46,9 @@ export module Components {
       this._colorScale = colorScale;
       this._redrawCallback = (scale) => this.redraw();
       this._colorScale.onUpdate(this._redrawCallback);
-
+      this._formatter = Formatters.identity();
       this.xAlignment("right").yAlignment("top");
-      this.comparator((a: string, b: string) => this._colorScale.domain().indexOf(a) - this._colorScale.domain().indexOf(b));
+      this.comparator((a: string, b: string) => 0);
```
Comment (bluong) [inline src/components/legend.ts]: Why this change?

File: test/components/legendTests.ts [hunk 24]
```diff
@@ -254,6 +254,43 @@ describe("Legend", () => {
     svg.remove();
   });
 
+  it("can set formatter and change how entry labels are displayed", () => {
```
Comment (bluong) [inline test/components/legendTests.ts]: "can set formatter to change..."

File: test/components/legendTests.ts [hunk 25]
```diff
@@ -254,6 +254,43 @@ describe("Legend", () => {
...
+    legendRows.each(function(d, i){
+      let expectText = formatter(legend.colorScale().domain()[i]);
+      assert.strictEqual(d3.select(this).select("text").text(), expectText,
+        `formatter output ${expectText} should be displayed`);
+    });
+    svg.remove();
+  });
+
+  it("can get formatter of the legend using formatter()", () => {
+    let formatter = (id: string) => String.fromCharCode(id.charCodeAt(0) + 1);
```
Comment (bluong) [inline test/components/legendTests.ts]: I personally like something even simpler like `(id: string) => \`${id}foo\``

The problem that I have with this formatter is that there are 3 things happening in order to achieve the formatted value, which is just needless overhead if people needed to decipher what happens if a test were to fail with this formatter

File: src/components/legend.ts (hunk 23 above)
Comment (yunhaolucky) [inline src/components/legend.ts]: The parameters of the old `comparator` are domain names. It fails when `a` and `b` are displayed texts. The new comparator will not sort displayed texts, whose order is the same as the order of the index of domain names.

File: test/components/legendTests.ts [hunk 26]
```diff
@@ -254,6 +254,43 @@ describe("Legend", () => {
...
 
+  it("can set formatter to change how entry labels are displayed", () => {
+    color.domain(["A", "B", "C"]);
+    let formatter = (id: string) => `${id}foo`;
+    legend.formatter(formatter);
+    legend.renderTo(svg);
+    let legendRows = svg.selectAll(entrySelector);
+    assert.operator(legendRows.size(), ">=", 1, "There is at least one entry in the test");
+    legendRows.each(function(d, i){
+      let expectText = formatter(legend.colorScale().domain()[i]);
```
Comment (bluong) [inline test/components/legendTests.ts]: Sorry for the late catch here.

`expectedText` please

**Comment by crmorford** [comment]
http://jsfiddle.net/cmf3ut7g/5/
http://jsfiddle.net/cmf3ut7g/6/
